src/constraints/RebarBondSlipConstraint.C

```diff
@@ -0,0 +1,348 @@
+      // if (_current_node->id() == 144)
+      std::cout << "Calculating bondstress for Case III"
+                << ".\n";
+
+    _bond_stress = (slip - plastic_slip_min) * bond_slip.bondstress_min /
+                   (bond_slip.slip_min - plastic_slip_min);
+  }
+  else
+    _bond_stress = _frictional_bondstress;
+
+  if (_debug)
+    // if (_current_node->id() == 144)
+    std::cout << "Bondstress = " << _bond_stress << "\n";
+
+  Real bond_force = 2.0 * libMesh::pi * _bar_radius * _current_elem_volume * _bond_stress;
+
+  RealVectorValue constraint_force_axial = bond_force * _slave_tangent;
+  RealVectorValue constraint_force_normal = _penalty * slip_normal;
+
+  _constraint_residual = constraint_force_axial + constraint_force_normal;
+
+  if (_debug)
+  // if (_current_node->id() == 144)
+  {
+    std::cout << "Constraint Residual Axial = " << constraint_force_axial << "\n";
+    std::cout << "Constraint Residual Normal = " << constraint_force_normal << "\n";
+    std::cout << "Constraint Residual = " << _constraint_residual << "\n";
+  }
+
+  bond_slip.bondstress_min = std::min(bond_slip.bondstress_min_old, _bond_stress);
+  bond_slip.bondstress_max = std::max(bond_slip.bondstress_max_old, _bond_stress);
+
+  if (_fe_problem.converged())
+  {
+    _bondslip[node->id()].slip_min = bond_slip.slip_min;
+    _bondslip[node->id()].slip_max = bond_slip.slip_max;
+    _bondslip[node->id()].bondstress_min = bond_slip.bondstress_min;
+    _bondslip[node->id()].bondstress_max = bond_slip.bondstress_max;
+  }
+}
+
+Real
+RebarBondSlipConstraint::computeQpResidual(Moose::ConstraintType type)
+{
+  Real resid = _constraint_residual(_component);
+
+  switch (type)
+  {
+    case Moose::Slave:
+      return _test_slave[_i][_qp] * resid;
+
+    case Moose::Master:
+      return _test_master[_i][_qp] * -resid;
+  }
+
+  return 0.0;
+}
+
+Real
+RebarBondSlipConstraint::computeQpJacobian(Moose::ConstraintJacobianType type)
+{
+  switch (type)
+  {
+    case Moose::SlaveSlave:
+      return _phi_slave[_j][_qp] * _penalty * _test_slave[_i][_qp] *
+             (1.0 - _slave_tangent(_component) * _slave_tangent(_component));
+
+    case Moose::SlaveMaster:
+      return -_phi_master[_j][_qp] * _penalty * _test_slave[_i][_qp] *
+             (1.0 - _slave_tangent(_component) * _slave_tangent(_component));
+
+    case Moose::MasterSlave:
+      return -_test_master[_i][_qp] * _penalty * _phi_slave[_j][_qp] *
+             (1.0 - _slave_tangent(_component) * _slave_tangent(_component));
+
+    case Moose::MasterMaster:
+      return _test_master[_i][_qp] * _penalty * _phi_master[_j][_qp] *
+             (1.0 - _slave_tangent(_component) * _slave_tangent(_component));
+
+    default:
+      mooseError("Unsupported type");
+      break;
+  }
+  return 0.0;
+}
+
+Real
+RebarBondSlipConstraint::computeQpOffDiagJacobian(Moose::ConstraintJacobianType /*type*/,
+                                                  unsigned int /*jvar*/)
+{
+  return 0.0;
```

Comment:
It's going to be important to implement this for anything off-axis.

Reply:
By this I mean that if the bar is oriented in a direction off the Cartesian axes, this will be nonzero.  You can ignore this for the first cut, but this will be very important to get right.